R/schedule.R

```diff
@@ -21,132 +21,109 @@ get_tune_schedule <- function(wflow, param, grid) {
 
-	# ------------------------------------------------------------------------------
-	# Add info an any postprocessing parameters
+	schedule
+}
 
-	sched <- sched %>%
-		dplyr::mutate(
-			predict_stage = purrr::map(
-				predict_stage,
-				~.x %>% dplyr::group_nest(!!!symbs$sub, .key = "post_stage")
-			)
-		)
+schedule_stages <- function(grid, param_info, wflow) {
+	# schedule preprocessing stage and push the rest into a nested tibble
+	param_pre_stage <- param_info %>% 
+    filter(source == "recipe") %>% 
+    pull(id)
+	schedule <- grid %>% 
+    tidyr::nest(.by = all_of(param_pre_stage), .key = "model_stage")
+
+	# schedule next stages recursively
+	schedule %>% 
+		mutate(
+      model_stage = 
+        purrr::map(
+					model_stage,
+					schedule_model_stage_i, 
+					param_info = param_info,
+					wflow = wflow
+				)
+    )
+}
 
-	# ------------------------------------------------------------------------------
-	# Merge in submodel fit value (if any)
+schedule_model_stage_i <- function(model_stage, param_info, wflow){
+  model_param <- param_info %>% 
+    filter(source == "model_spec") %>% 
+    pull(id)
+  non_submodel_param <- param_info %>% 
+    filter(source == "model_spec" & !has_submodel) %>% 
+    pull(id)
+  
+  # schedule model parameters
+  schedule <- min_model_grid(model_stage, model_param, wflow)
+
+  # push remaining paramters into the next stage
+  next_stage <- model_stage %>% 
+    tidyr::nest(.by = all_of(non_submodel_param), .key = "predict_stage")
+
+  schedule <- schedule %>% 
+    dplyr::left_join(next_stage, by = all_of(non_submodel_param))
+
+	# schedule next stages recursively
+	schedule %>% 
+		mutate(
+      predict_stage = 
+        purrr::map(predict_stage, schedule_predict_stage_i, param_info = param_info)
+    )
+}
 
-	loop_names <- names(sched)[names(sched) != "predict_stage"]
-	if (length(loop_names) > 0) {
-		# Using `by = character()` to perform a cross join was deprecated
-		sched <- dplyr::full_join(sched, first_loop_info, by = loop_names)
-	}
+min_model_grid <- function(grid, model_param, wflow){
+  # work on only the model parameters
+  model_grid <- grid %>% 
+    select(all_of(model_param)) %>% 
+    dplyr::distinct()
+
+  min_grid(
+    extract_spec_parsnip(wflow),
+    model_grid
+  ) %>% 
+    select(all_of(model_param))
+}
 
-	# ------------------------------------------------------------------------------
-	# Now collapse over the preprocessor for conditional execution
+schedule_predict_stage_i <- function(predict_stage, param_info) {
+  submodel_param <- param_info %>% 
+    filter(source == "model_spec" & has_submodel) %>% 
+    pull(id)
 
-	sched <- sched %>% dplyr::group_nest(!!!symbs$pre, .key = "model_stage")
+  predict_stage %>% 
+    tidyr::nest(.by = all_of(submodel_param), .key = "post_stage")
+}
 
-	# ------------------------------------------------------------------------------
+get_param_info <- function(wflow) {
+  param_info <- tune_args(wflow) %>% 
```

Comment:
Using `tune_args()` here instead of a parameter set object, due to considerations I've put in https://github.com/tidymodels/tune/pull/974#issuecomment-2590393148